mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java

```diff
@@ -0,0 +1,124 @@
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.transport.mailets;
+
+import java.util.Optional;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.core.MailAddress;
+import org.apache.mailet.Mail;
+import org.apache.mailet.Mailet;
+import org.apache.mailet.MailetException;
+import org.apache.mailet.base.GenericMailet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+
+/**
+ * <p>Collects the sender and the recipients of a message and store them as JSON in a specified message attribute.</p>
+ * <p>Here is the JSON format:</p>
+ * <pre><code>
+ * {
+ *   userEmail : sender@james.org, 
+ *   emails : [ to@james.org, cc@james.org]
+ * }
+ * </code></pre>
+ * 
+ * <p>Sample configuration:</p>
+ *
+ * <pre><code>
+ * &lt;mailet match="All" class="ContactExtractor"&gt;
+ *   &lt;attribute&gt;ExtractedContacts&lt;/attribute&gt;
+ * &lt;/mailet&gt;
+ * </code></pre>
+ */
+public class ContactExtractor extends GenericMailet implements Mailet {
+
+    public static final String ATTRIBUTE = "attribute";
+    private static final Logger LOGGER = LoggerFactory.getLogger(ContactExtractor.class);
+
+    @VisibleForTesting ObjectMapper objectMapper;
+    private String extractAttributeTo;
+
+    @Override
+    public void init() throws MessagingException {
+        extractAttributeTo = getInitParameter(ATTRIBUTE);
+        
+        if (Strings.isNullOrEmpty(extractAttributeTo)) {
+            throw new MailetException("No value for " + ATTRIBUTE + " parameter was provided.");
+        }
+        objectMapper = new ObjectMapper().registerModule(new Jdk8Module());
+    }
+
+    @Override
+    public String getMailetInfo() {
+        return "ContactExtractor Mailet" ;
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        try {
+            LOGGER.debug("extracting recipients");
+            Optional<String> payload = buildMessage(mail);
+            LOGGER.debug("payload : {}", payload);
+            payload.ifPresent(x -> mail.setAttribute(extractAttributeTo, x));
+        } catch (Exception e) {
+            LOGGER.error("Error while extracting contacts", e);
+        }
+    }
+
+    private Optional<String> buildMessage(Mail mail) {
```

Comment:
buildMessage is not a good name in this context.
In fact the object Message is bad names, there are already too much messages in a mail server :)

Reply:
I don't agree, same name in different contexts are ok. It's why it's a nested object.

Reply:
nested object or not, when you read buildMessage(mail) you don't except at first sight to handle something other than mails.